sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("lastUpdatedDateTimeUtc"), tm->name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->error_timestamp));
+    }
+    else if (tm->token.kind == AZ_JSON_TOKEN_OBJECT_START)
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_skip_children(jp, tm->token));
+    }
+    else
+    {
+      (void)az_iot_provisioning_client_payload_error_code_parse(tm, out_state);
+    }
+  }
+
+  return AZ_OK;
+}
+
+AZ_INLINE az_result az_iot_provisioning_client_parse_payload(
+    az_span received_payload,
+    az_iot_provisioning_client_register_response* out_response)
+{
+  // Parse the payload:
+  az_json_parser jp;
+  az_json_token_member tm;
+  bool found_operation_id = false;
+  bool found_registration_state = false;
+  bool found_error = false;
+
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, received_payload));
+  AZ_RETURN_IF_FAILED(az_json_parser_parse_token(&jp, &tm.token));
+  if (tm.token.kind != AZ_JSON_TOKEN_OBJECT_START)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  out_response->registration_information = registration_state_default();
+
+  while (az_succeeded(az_json_parser_parse_token_member(&jp, &tm)))
+  {
+    if (az_span_is_content_equal(AZ_SPAN_FROM_STR("operationId"), tm.name))
+    {
+      found_operation_id = true;
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm.token, &out_response->operation_id));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("status"), tm.name))
+    {
+      found_registration_state = true;
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm.token, &out_response->registration_state));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("registrationState"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_iot_provisioning_client_payload_registration_state_parse(
+          &jp, &tm, &out_response->registration_information));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("trackingId"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_tracking_id));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("message"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_message));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("timestampUtc"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_timestamp));
+    }
+    else if (tm.token.kind == AZ_JSON_TOKEN_OBJECT_START)
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_skip_children(&jp, tm.token));
+    }
+    else if (az_succeeded(az_iot_provisioning_client_payload_error_code_parse(
+                 &tm, &out_response->registration_information)))
+    {
+      found_error = true;
+    }
+
+    // else ignore token.
```

Comment:
Should this be considered an error condition?

Reply:
Definitely not: it would break if it is (why I've added the comment - to make it clear that else, we'll ignore the token).
In parallel, I've opened a task to log all the incoming payloads that we parse. If we will see customer issues we can inspect the logs to figure out the correct repro.

Reply:
Do we expect having more tokens/info within the JSON than precisely what is required? OR is this about ignoring user defined component of the JSON payload?

If so, I think we may want to verify and "skip" past the "payload" JSON property (and its content), but still error on any other arbitrary token.

Reply:
We do expect more _valid_ tokens and we want to skip those until we read the ones we care about.

I think it's good to parse only the items we're interested in (for some reason the service is too verbose here). We're not validating the server for items that we don't expose/use or items that won't be useful for applications (at least to our knowledge today).